fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java

```diff
@@ -138,100 +137,134 @@ public LuceneIndexMaintainer(@Nonnull final IndexMaintainerState state, @Nonnull
+            }
+            return evaluatedKeys;
+        }
+        return Lists.newArrayList();
+    }
+
+    private void insertField(DocumentEntry entry, final Document document) {
+        switch (entry.expression.getType()) {
+            case INT:
+                // Todo: figure out how to expand functionality to include storage, sorting etc.
+                document.add(new IntPoint(entry.fieldName, (Integer)entry.value));
+                break;
+            case STRING:
+                document.add(new TextField(entry.fieldName, entry.value == null ? "" : (String)entry.value, entry.expression.isStored() ? Field.Store.YES : Field.Store.NO));
+                break;
+            case LONG:
+                document.add(new LongPoint(entry.fieldName, (long)entry.value));
+                break;
+            default:
+                throw new RecordCoreArgumentException("Invalid type for lucene index field", "type", entry.expression.getType());
+        }
+    }
+
     @Nonnull
     @Override
     public <M extends Message> CompletableFuture<Void> update(@Nullable FDBIndexableRecord<M> oldRecord,
                                                               @Nullable FDBIndexableRecord<M> newRecord) {
         LOG.trace("update oldRecord={}, newRecord{}", oldRecord, newRecord);
-        return super.update(oldRecord, newRecord);
-    }
 
-    @Nonnull
-    @Override
-    protected <M extends Message> CompletableFuture<Void> updateIndexKeys(@Nonnull final FDBIndexableRecord<M> savedRecord,
-                                                                          final boolean remove,
-                                                                          @Nonnull final List<IndexEntry> indexEntries) {
-        LOG.trace("updateIndexKeys savedRecord={}, remove=={}, indexEntries={}", savedRecord, remove, indexEntries);
+        List<DocumentEntry> oldRecordFields = Lists.newArrayList();
+        KeyExpression root = state.index.getRootExpression();
+        List<DocumentEntry> newRecordFields = Lists.newArrayList();
 
-        if (indexEntries.isEmpty()) {
-            return AsyncUtil.DONE;
+        // evaluate fields from records provided
+        if (newRecord != null) {
+            newRecordFields.addAll(getFields(root, newRecord, newRecord.getRecord()));
         }
+        if (oldRecord != null) {
+            oldRecordFields.addAll(getFields(root, oldRecord, oldRecord.getRecord()));
+        }
+
+        // evaluate which fields are changed
+        //Not sure if we want to allow deletion of fields. Or if the updated record will contain all fields not
+        // just the updated ones. Assuming here that the record will be complete if being updated as previously.
+        List<DocumentEntry> updatedFields = newRecordFields;
+        updatedFields.removeAll(oldRecordFields);
```

Comment:
This assumes that `DocumentEntry` has a meaningful (non-reference) `equals`, which it does not right now.